packages/prettier-plugin-java/src/printers/types-values-and-variables.js

```diff
@@ -1,81 +1,194 @@
 "use strict";
 /* eslint-disable no-unused-vars */
 
+const {
+  concat,
+  join,
+  line,
+  ifBreak,
+  group,
+  indent,
+  dedent
+} = require("prettier").doc.builders;
+const { rejectAndJoin } = require("./printer-utils");
+
 class TypesValuesAndVariablesPrettierVisitor {
   primitiveType(ctx) {
-    return "primitiveType";
+    const annotations = this.mapVisit(ctx.annotation);
+    let type = null;
+    if (ctx.numericType) {
+      type = this.visit(ctx.numericType);
+    } else {
+      type = this.getSingle(ctx).image;
+    }
+
+    return rejectAndJoin(" ", [join(" ", annotations), type]);
   }
 
   numericType(ctx) {
-    return "numericType";
+    return this.visitSingle(ctx);
   }
 
   integralType(ctx) {
-    return "integralType";
+    return this.getSingle(ctx).image;
   }
 
   floatingPointType(ctx) {
-    return "floatingPointType";
+    return this.getSingle(ctx).image;
   }
 
   referenceType(ctx) {
-    return "referenceType";
+    const annotations = this.mapVisit(ctx.annotation);
+
+    const type = ctx.primitiveType
+      ? this.visit(ctx.primitiveType)
+      : this.visit(ctx.classOrInterfaceType);
+
+    const dims = this.visit(ctx.dims);
+
+    return rejectAndJoin("", [join(" ", annotations), " ", type, dims]);
   }
 
   classOrInterfaceType(ctx) {
-    return "classOrInterfaceType";
+    return this.visitSingle(ctx);
   }
 
   classType(ctx) {
-    return "classType";
+    const dots = ctx.Dots;
+
+    if (dots) {
```

Comment:
maybe if you sort **everything** (except dots) by the offset first you would have an easy to work with structure?
it would look very much like the grammar.
> (annotation)* identifier (typeArguments)?

Reply:
I thought about it. What could be difficult is that we can only have information on tokens' offset : it would be more difficult to know the offset of a rule, unless we start exploring its children...

Reply:
Yes but exploring the first element of annotation or typeArguments to determine their offset should be trivial, we do not need a generic solution for offset range for any and all CSTNodes.